File: utils/git/log.py
import subprocess
from rich.table import Table
from rich import box

class GitLog:
    """
    Visualization tools for Git history: log and reflog.
    """

    def show_log(self, count: int = 10):
        """
        Retrieve and render the commit history in a formatted Rich table.
        """
        try:
            result = subprocess.run(
                ["git", "log", f"-n", str(count), "--pretty=format:%h|%ad|%an|%s", "--date=format:%Y-%m-%d %H:%M"],
                capture_output=True, text=True, check=True
            )
            
            table = Table(title="Git History", box=box.ROUNDED, border_style="green")
            table.add_column("Hash", style="dim green", no_wrap=True)
            table.add_column("Date & Time", style="green")
            table.add_column("Author", style="yellow")
            table.add_column("Message", style="white")

            for line in result.stdout.strip().split("\n"):
                if "|" in line:
                    table.add_row(*line.split("|"))
            
            self.console.print(table)
        except Exception:
            self.console.print("[bold red]Error:[/] Could not fetch log.")

    def show_reflog(self, count: int = 10):
        """
        Retrieve and render the Git reflog in a formatted Rich table.
        """
        try:
            result = subprocess.run(
                ["git", "reflog", f"-n", str(count), "--pretty=format:%h|%ad|%gs", "--date=format:%Y-%m-%d %H:%M"],
                capture_output=True, text=True, check=True
            )
            
            table = Table(title="Reflog (Recovery)", box=box.ROUNDED, border_style="magenta")
            table.add_column("Index", style="dim white", justify="right")
            table.add_column("Hash", style="dim magenta", no_wrap=True)
            table.add_column("Time", style="green")
            table.add_column("Operation", style="white")

            for i, line in enumerate(result.stdout.strip().split("\n")):
                if "|" in line:
                    table.add_row(f"HEAD@{{{i}}}", *line.split("|"))
            
            self.console.print(table)
        except Exception:
            self.console.print("[bold red]Error:[/] Could not fetch reflog.")

File: internal/git/log.go
package git

import (
	"fmt"
	"strconv"
	"strings"
)

type LogEntry struct {
	Hash    string
	Time    string
	Author  string
	Message string
}

type ReflogEntry struct {
	Ref     string
	Hash    string
	Time    string
	Message string
}

func (g *GitCore) ShowLog(count int) ([]LogEntry, error) {
	count = normalizeCount(count)

	stdout, stderr, err := g.RunCommand(
		"log",
		"-n",
		strconv.Itoa(count),
		"--pretty=format:%h|%ad|%an|%s",
		"--date=format:%Y-%m-%d %H:%M",
	)
	if err != nil {
		if stderr != "" {
			return nil, fmt.Errorf("git log failed: %s", stderr)
		}
		return nil, fmt.Errorf("git log failed: %w", err)
	}

	if strings.TrimSpace(stdout) == "" {
		return []LogEntry{}, nil
	}

	lines := strings.Split(stdout, "\n")
	entries := make([]LogEntry, 0, len(lines))
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		parts := strings.SplitN(line, "|", 4)
		if len(parts) != 4 {
			return nil, fmt.Errorf("failed to parse git log line: %q", line)
		}

		entries = append(entries, LogEntry{
			Hash:    strings.TrimSpace(parts[0]),
			Time:    strings.TrimSpace(parts[1]),
			Author:  strings.TrimSpace(parts[2]),
			Message: strings.TrimSpace(parts[3]),
		})
	}

	return entries, nil
}

func (g *GitCore) ShowReflog(count int) ([]ReflogEntry, error) {
	count = normalizeCount(count)

	stdout, stderr, err := g.RunCommand(
		"reflog",
		"-n",
		strconv.Itoa(count),
		"--pretty=format:%h|%ad|%gs",
		"--date=format:%Y-%m-%d %H:%M",
	)
	if err != nil {
		if stderr != "" {
			return nil, fmt.Errorf("git reflog failed: %s", stderr)
		}
		return nil, fmt.Errorf("git reflog failed: %w", err)
	}

	if strings.TrimSpace(stdout) == "" {
		return []ReflogEntry{}, nil
	}

	lines := strings.Split(stdout, "\n")
	entries := make([]ReflogEntry, 0, len(lines))
	for i, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}

		parts := strings.SplitN(line, "|", 3)
		if len(parts) != 3 {
			return nil, fmt.Errorf("failed to parse git reflog line: %q", line)
		}

		entries = append(entries, ReflogEntry{
			Ref:     fmt.Sprintf("HEAD@{%d}", i),
			Hash:    strings.TrimSpace(parts[0]),
			Time:    strings.TrimSpace(parts[1]),
			Message: strings.TrimSpace(parts[2]),
		})
	}

	return entries, nil
}

func normalizeCount(count int) int {
	if count <= 0 {
		return 10
	}
	return count
}
